### PR TITLE
Fix KerasMetricCallback prediction with generate() and inference of column names

### PR DESCRIPTION
**Pull request by Rocketknight1**
A number of fixes to the Keras metric callback based on testing with the notebooks. I think I've caught all the implementation differences, and in most cases we can now pass exactly the same `compute_metrics` function to this callback as to `Trainer`.

I'm still in the process of overhauling the notebooks, so I might need to make a few more tweaks to this PR if I encounter any other problems!

### CODE REVIEW

File: src/transformers/keras_callbacks.py [hunk 1]
```diff
@@ -169,9 +192,13 @@ def on_epoch_end(self, epoch, logs=None):
                 predictions = self.model.generate(generation_inputs, attention_mask=attention_mask)
             else:
                 predictions = self.model.predict(batch)
-            predictions = dict(predictions)
+            if isinstance(predictions, dict):
+                # BatchEncoding is a subclass of dict
+                predictions = dict(predictions)
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:197]: I need a comment here that will stop me from thinking you're gone crazy ;-)

(I think I know why you're doing that, but not quite sure.)

File: src/transformers/keras_callbacks.py [hunk 2]
```diff
@@ -160,7 +183,7 @@ def on_epoch_end(self, epoch, logs=None):
                 labels = None
             if self.predict_with_generate:
                 if isinstance(batch, dict):
-                    generation_inputs = batch[self.main_input_name]
+                    generation_inputs = batch["input_ids"]
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:186]: This seems like a bad change.

File: src/transformers/keras_callbacks.py [hunk 3]
```diff
@@ -123,32 +114,64 @@ def __init__(
...
+        num_samples = sum([batch.shape[0] for batch in batches])
+        output = np.full_like(
+            batches[0], fill_value=padding_index, shape=[num_samples, max_len] + list(batches[0].shape[2:])
+        )
+        # i keeps track of which part of the concatenated array we're writing the next batch to
+        i = 0
+        for batch in batches:
+            output[i : i + len(batch), : batch.shape[1]] = batch
+            i += len(batch)
+        return output
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:146]: Nice!

**Review by sgugger — APPROVED** [review]
Thanks for fixing!

File: src/transformers/keras_callbacks.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py:197]: The answer is that Keras does some *very* weird things if you pass it a `BatchEncoding`, or other dict-subclass. It has to be a `dict`, for some reason. But yes, I should explain it in the comments!

File: src/transformers/keras_callbacks.py (hunk 2 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py:186]: Ah, sorry, this was a change I made in testing! I'll put the `main_input_name` code back, you're right.

Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py]: Fixed!

**Review by gante — APPROVED** [review]
👍
